### PR TITLE
Dataframe join_on method

### LINKED ISSUE #1254: Support non-equi join (e.g. `ON` clause) in Dataframe API
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Users should be able to use the Dataframe API to expression the following SQL queries:

```
SELECT * FROM fact JOIN dim ON fact.x != dim.y;
SELECT * FROM fact JOIN dim ON fact.x > dim.y;
SELECT * FROM fact JOIN dim ON fact.x <= dim.y;
SELECT * FROM fact JOIN dim ON fact.x <> dim.y;
```

**Describe the solution you'd like**

Potential solution would be changing the join method to take a list of expressions instead of `left_cols` and `right_cols`.

**Additional context**
Extension of https://github.com/apache/arrow-datafusion/pull/1135

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #1254

# Rationale for this change

# What changes are included in this PR?

New method for DataFrame `join_on` allowing user to pass in arbitrary `Expr`'s which are AND'ed together to form the `ON` condition.

Also fix to DataFrame join to enforce ambiguity check, like how was done by SQL planner

# Are these changes tested?

New unit test

# Are there any user-facing changes?

New method in DataFrame, doc updated

### CODE REVIEW

File: datafusion/expr/src/logical_plan/builder.rs
Comment (Jefffrey) [inline datafusion/expr/src/logical_plan/builder.rs:526]: related to https://github.com/apache/arrow-datafusion/issues/4196

fix bug where you could do dataframe join with ambiguous column for the filter expr

instead of having the check done in both DataFrame join api and SQL planner join mod, unify by having check done inside the logical plan builder

this is technically an unrelated fix to the actual issue, so i can extract into separate issue if needed

File: datafusion/core/src/dataframe.rs
Comment (alamb) [inline datafusion/core/src/dataframe.rs:397]: 👍  LGTM

File: datafusion/core/src/dataframe.rs
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1104]: Would it be possible here to also add an equality predicate to demonstrate they are automatically recognized as equi preds?

Perhaps something like

```suggestion
                col("a.c2").eq(col("b.c2")),
```

File: datafusion/expr/src/logical_plan/builder.rs
Comment (alamb) [reply datafusion/expr/src/logical_plan/builder.rs:526]: I think it is fine to include in this PR as long as it also has a test (for ambiguity check using the DataFrame API)

**Review by alamb — COMMENTED** [review]
Thanks @Jefffrey  -- the code looks great and I have just a few small comments on tests

File: datafusion/expr/src/logical_plan/builder.rs
Comment (Jefffrey) [reply datafusion/expr/src/logical_plan/builder.rs:526]: test added

**Comment by liukun4515** [comment]
I want to take a look this PR tomorrow. @alamb

File: docs/source/user-guide/dataframe.md
Comment (alamb) [inline docs/source/user-guide/dataframe.md:71]: ❤️

**Review by alamb — APPROVED** [review]
LGTM -- thanks @Jefffrey  -- let's wait for @liukun4515 to review before merging

**Review by liukun4515 — APPROVED** [review]
LGTM

**Pull request merged**
